aten/src/ATen/core/type.cpp

```diff
@@ -747,6 +751,74 @@ TupleTypePtr TupleType::createNamed(
       field_types, qualName, schema)); // NOLINT(modernize-make-shared)
 }
 
+void flatten_union(TypePtr& type, std::vector<TypePtr>& res) {
+  if (type->kind() == UnionType::Kind) {
+    for (auto inner_type : type->expect<UnionType>()->types()) {
+      flatten_union(inner_type, res);
+    }
+  }
+  else {
+    res.emplace_back(type);
+  }
+}
+
+UnionType::UnionType(std::vector<TypePtr> types) :
+  Type(TypeKind::UnionType) {
+  // Flatten Unions of Unions while populating `types_`
+  for (auto type : types) {
+    flatten_union(type, types_);
+  }
+  // We want the elements to be sorted so we can easily compare two UnionType
+  // objects for equality in the future
+  std::sort(types_.begin(), types_.end(),
+  [](const TypePtr a, const TypePtr b) -> bool {
+    return a->kind() > b->kind();
+  });
+  // Filter out duplicate types
+  types_.erase(std::unique(types_.begin(), types_.end()), types_.end());
+  // Check if this Union allows None
+  can_hold_none_ = std::any_of(types_.begin(), types_.end(),
+    [](const TypePtr type) { return type->kind() == NoneType::Kind; });
+}
+
+bool UnionType::isSubtypeOfExt(const TypePtr& rhs_, std::ostream* why_not) const {
+  if (auto union_rhs = rhs_->cast<UnionType>()) {
+    return std::all_of(this->types_.begin(), this->types_.end(),
```

Comment:
I think this can be slightly optimized. Right now complexity is `O(n^2)` where `n` is number of types. Since all types within `union` are sorted, I think you can search both lists of types linearly to find if one is subset of the other.

Reply:
I would normally just compare the two vectors, since I think `std::vector` implements `operator=`, but I did it this way because I'm worried about the subtyping relationships. What will happen if we have a class `A` and a subclass `B` and we compare `x = Union[A, ...]` with `y = Union[B, ...]`? Those `A` and `B` might not have the same index in their respective Unions, but we should still say that `y` <: `x`